src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs

```diff
@@ -196,6 +207,77 @@ static ImmutableArray<AssetsFileLogMessage> ParseLogMessages(LockFile lockFile,
             }
 
             return builder.ToImmutable();
+
+            Dictionary<string, LogLevel> BuildLevelByLibrary()
+            {
+                Dictionary<string, HashSet<string>> ancestorsByLibrary = new(StringComparer.OrdinalIgnoreCase);
+
+                // Build a map from child-to-ancestors
+                foreach (LockFileTargetLibrary lockFileLibrary in lockFileTarget.Libraries)
+                {
+                    string? parentId = lockFileLibrary.Name;
+
+                    if (parentId is null)
+                    {
+                        continue;
+                    }
+
+                    foreach (var dep in lockFileLibrary.Dependencies)
+                    {
+                        string childId = dep.Id;
+
+                        if (!ancestorsByLibrary.TryGetValue(childId, out HashSet<string>? ancestors))
+                        {
+                            ancestors = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                            ancestorsByLibrary.Add(childId, ancestors);
+                        }
+
+                        ancestors.Add(parentId);
+                    }
+                }
+
+                // Walk up the tree from each log message's library, propagating the highest level found.
+                Dictionary<string, LogLevel> levelByLibrary = new(StringComparer.OrdinalIgnoreCase);
+
+                foreach (AssetsFileLogMessage message in logMessages)
+                {
+                    Integrate(message.LibraryName, message.Level, visited: []);
+                }
+
+                return levelByLibrary;
+
+                void Integrate(string id, LogLevel level, HashSet<string> visited)
+                {
+                    if (!visited.Add(id))
+                    {
+                        // Avoid infinite recursion.
+                        return;
+                    }
+
+                    if (!levelByLibrary.TryGetValue(id, out LogLevel currentLevel))
+                    {
+                        // No level yet, so set it.
+                        levelByLibrary[id] = level;
+                    }
+                    else
+                    {
+                        // Higher level is more severe. If we already have a higher level, don't change it.
+                        // Also, if we already have a level, assume we have propagated it to all ancestors and can return.
+                        if (currentLevel >= level)
+                        {
+                            return;
+                        }
+                    }
+
+                    if (ancestorsByLibrary.TryGetValue(id, out HashSet<string>? ancestors))
+                    {
+                        foreach (string ancestor in ancestors)
```

Comment:
When we create the ancestors set,  could we have them sorted (have the largest one first)? Just wondering if that that might allow us to select the highest loglevel first and avoid more recursion

Reply:
Interesting idea. The only way I could see this making a difference is sorting the top-level log messages by descending severity, to increase the chance that later walks up the graph short-circuit. I don't expect it'd have a lot of impact, but I can try it.